Allow custom orders to affect face quadrature in addition to vol

rwcarlsen opened the pull request:
Fixes #15472.  @YaqiWang - take a look at this and let me know how close it is to what you need.

rwcarlsen committed:
allow custom face quadrature rules

This allows face quadrature rules to be together specified with the
recently added block-custom volume quadrature rules.

Fixes #15472.

rwcarlsen committed:
include checks for face quadrature in the test, regold test

YaqiWang commented on framework/include/base/Assembly.h:
```diff
@@ -1804,12 +1812,21 @@ class Assembly
+  /// the subdomain that has the highest specified quadrature order.  So when
+  /// you need to access a face quadrature rule, you should retrieve it via this
+  /// function.
+  QBase * qruleFace(const Elem * face);
```
We typically do not have Elem for a face, but instead refer to a face by something like `(const Elem * elem, unsigned int side)`.

rwcarlsen replied:
This was in order to use the existing node block ids function on moosemesh.  We may want to come up with a better alternative to that function though.

YaqiWang commented on framework/src/base/Assembly.C:
```diff
@@ -1801,6 +1816,28 @@ Assembly::reinitFVFace(const FaceInfo & fi)
+  std::set<SubdomainID> blocks;
+  for (subdomain_id_type i = 0; i < face->n_nodes(); i++)
+  {
+    auto & ids = _mesh.getNodeBlockIds(*(face->node_ptr(i)));
```
If we pass in volumetric element and side, we do not need to call this costly function. I do not think we need to check all elements touching this side. We just need to check its two neighbors.

rwcarlsen replied:
Yeah.  This isn't a very good way - I agree.  I'll rework it to look at the neighboring elem instead of nodes.  This was just my lazy programmer approach trying to (re)use existing code.

YaqiWang commented on test/tests/quadrature/order/block-order.i:
```diff
@@ -10,20 +10,27 @@
   []
 
-  [left]
+  [bottom]
```
I will push in a test to make sure this change works for Griffin.

YaqiWang committed:
a test for face quadrature order #15472

YaqiWang commented:
@rwcarlsen my test can be found here: 423a761.

rwcarlsen commented:
Okay - @YaqiWang - this has been cleaned up to use neighbor pointers.

YaqiWang commented on framework/src/base/Assembly.C:
```diff
@@ -1846,8 +1879,7 @@ Assembly::reinit(const Elem * elem, unsigned int side, const std::vector<Point>
   // Make sure the qrule is the right one
-  if (_current_qrule_face != _current_qrule_arbitrary_face)
-    setFaceQRule(_current_qrule_arbitrary_face, elem_dimension);
+  setFaceQRule(_current_qrule_arbitrary_face, elem_dimension);
```
why is if statement removed?

rwcarlsen replied:
I had been experimenting with moving that if check down into the setFaceQRule function, but it turns out there are apparently side effects of calling that function that cause other things to break if we _always_ perform that check instead of just whenever the caller wants to :-(  This was just an oversight after that failed attempt - I'll put this if check back in.

YaqiWang commented:
Can you cherry-pick my commit and add a test? One line in the source needs to be fixed as you have pointed out. Other than this and the above small question, looks good to me now.

rwcarlsen commented:
This PR extends the pre-existing test to check face qp counts in addition to cell/vol qp counts.  So it is tested.  Is there some particular/special issue that your material test is be checking for?

YaqiWang commented:
That test is a mimic of the use case in Griffin. It test all the sides surrounding a subdomain that has bumped quadratures. I guess I am worried if codes after Line 1833 in Assembly.C is visited or not. It also tests the call to that function. It does not hurt to add that into our test suites IMO.

rwcarlsen commented:
I'm fine adding that test - I'll patch it in.  FWIW - the existing test does confirm that the 1833+ code runs correctly.

rwcarlsen commented:
Turns out that materials test turned up a bug - somewhere materials aren't being resized properly for the changing number of qps per block with this.  Hopefully I can dig up and fix the issue today.

rwcarlsen committed:
fix+add Yaqi's material test

rwcarlsen committed:
remove wrong line from block order test input file

rwcarlsen commented:
Okay - this should be fixed and ready - at least ready enough to wait until CIVET comes back online :-(

rwcarlsen commented:
@YaqiWang - test this on some real problems if/when you get a chance and let me know if it does the trick.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/582185/) on 12b2f75 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15790/site)

This comment will be updated on new commits.

YaqiWang commented:
I am checking out this and changing Griffin to see if it is working properly. Will let you know soon.

loganharbour reviewed (dismissed):
This looks good.

I do have a question, though: in what case would we want to bump the volume quadrature but not the face? Shouldn't we just bump all regardless?

rwcarlsen commented:
Maybe?  The bumpVolQuadrature was created originally because custom face quadrature support wasn't available.  But maybe we never want to only bump vol?  I can't really answer that question - I'm not the one who has any current use cases. @YaqiWang?

loganharbour commented:
I'm in support of wrapping it up in `bumpQuadrature()` and bumping it all.

rwcarlsen commented:
Yeah - one consolidated function is probably best.  But let's make sure @YaqiWang doesn't need anything more granular.

YaqiWang commented:
What I need is to bump quadrature order in specific mesh blocks and the sides on those blocks including the boundary sides of them next to other surrounding blocks. It seems to me that yes, I only need `bumpQuadrature` as @loganharbour suggested. I am still compiling Griffin.

YaqiWang commented:
All relevant tests in Griffin pass except one... Does not make sense. I am looking into it.

YaqiWang commented:
That test only has one block and a material is calling `_fe_problem.bumpAllQRuleOrder` for that block. Can you think something could go wrong? The test shows random behavior, sometime nan, sometime finish properly but results into different solutions. I will message you on Slack.

YaqiWang commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -4686,6 +4686,18 @@ FEProblemBase::bumpVolumeQRuleOrder(Order order, SubdomainID block)
+  if (_displaced_problem)
+    _displaced_problem->bumpAllQRuleOrder(order, block);
+
+  updateMaxQps();
```
This function is problematic because it does not resize the default values in `Coupleable` and `MaterialPropertyInterface`.

YaqiWang reviewed (changes requested):
One Griffin test catches a potential bug in `updateMaxQps` function. I suggest to remove ` _coupleable_max_qps` and use the latest max Qps as well, otherwise the material calling `bumpAllQRuleOrder` calls `coupledValue` later will not see the change.

YaqiWang commented:
We may need to make `Coupleable::getDefaultValue` call a function in `FEProblemBase` to return the default `VariableValue`, i.e. let `FEProblemBase` own the values. Thus when we bump quadrature orders, we can resize those default values. We probably should do the same for default material properties. In this way, when we call `Coupleable::getDefaultValue` after the bump within the same object constructor, we can immediately see the new `MaxQp` and use it.

rwcarlsen commented:
Could someone else pick this up for me?  I'm currently working on some mildly urgent stuff for sebastian.  I'm not sure when I'll be able to pick this back up - maybe within a few weeks ish.

loganharbour commented:
I can get into it this week. @YaqiWang expressed that it wasn't an _immediate_ need, so that timing should work.

YaqiWang commented:
@loganharbour This is a reminder. I want this to be cleared in our list.

YaqiWang commented:
@jortensi and @friedmud , this the PR. @rwcarlsen or @loganharbour I do want this to be finished soon. It is not a good idea to keep it on the list. I'd say we finished 90% already.

loganharbour commented:
No

rwcarlsen commented:
@YaqiWang - working on this again - it seems to be running pretty clean - what was left that needed attention?  Other than maybe a consolidated `bumpQuadrature` function - it looks like you had a concern about resizing variables/materials for the bumped qp counts.  Was that a confirmed issue that we need to resolve here?  I don't see any failing tests - so we/I would need to construct a test that fails.  Can you comment/suggest on this?  Otherwise - without particular failing cases - I say we merge and address issues if/when they come up.

moosebuild commented:
Job [End User Experience (CentOS 8)](https://civet.inl.gov/job/653481/) on be235af : invalidated by @rwcarlsen

YaqiWang commented:
Please check my previous comments there. The PR cannot be merged as it is. Please check my custom_quadrature branch in my fork of Griffin. It contains one commit for Griffin to use this. One test is failing randomly as I mentioned.

rwcarlsen commented:
There are a few options to address the coupleable interface default variable value issue:

1. Do something like what Yaqi suggested - create a app-global collection of the default values and all the coupleable objects just have references/pointers to these values.  Then when we bump/mod quadrature rules and increase max qps, we can resize these default values and the objects will see the change.  I don't like this - it puts object specific code data and logic inside sim-global locations (e.g. fe problem).
2.  Create some sort of maxQpsChanged interface and call `updateMaxQps(new_max)` on every object implementing this interface whenever we bump/mod max qps.  This is okay-ish - but seems like way overkill.
3. Similar to 2 - we could have objects register themselves to get a maxQpsChanged callback.  Again - seems like a bit overkill.
4. Just modify coupleable objects to have like 50 or 100 for their maxqps w.r.t. default coupled var values.  Then we could add an error if maxqps goes above this number.  Or even add an FEProblem parameter that allows users to modify this "fixed" value.
5. Do something ridiculous like modify MooseArray to hold an FEProblem pointer and upon square bracket access - to dynamically make sure it is the right size, copying lower qp values to newly created ones if necessary.  This is more just for the sake of considering all possibilities.

Eventually, when all objects live in _the_ warehouse, it will be much easier/natural to just call maxQpsChanged on everything that needs it.  I'm inclined to do 4 - it is like a 1 line change - until we have a better, consolidated warehouse to work with.  @idaholab/moose-team ?

loganharbour commented:
I'm a fan of #4 - just use some default size that is large and produce a useful error if it's not large enough.

The fact that this size is object-specific (due to the need for constant variable values for a specific object) makes this a _very_ hard problem. I would prefer not to waste time on a solution that would require undoing a large workaround in the future. I think it is a much better move here to make the simple change, and enable improvement in the future (if needed). This is even more valid given the fact that we will be making a _large_ change in execution based on the warehouse, when a solution to this problem is more clear.

I'd also like to note that I agree with this:

> Do something like what Yaqi suggested - create a app-global collection of the default values and all the coupleable objects just have references/pointers to these values. Then when we bump/mod quadrature rules and increase max qps, we can resize these default values and the objects will see the change. I don't like this - it puts object specific code data and logic inside sim-global locations (e.g. fe problem).

FEProblem bloat right now is already quite awful. Keeping the addition of object-specific code in a global place that is already quite messy is a very sound design decision.

YaqiWang commented:
Ok, I think I am fine with 4 but I want you to add a check somewhere that when the maxqp is bigger than that hard-coded number, we error out. We possibly need to do the same for the default material properties.

rwcarlsen commented:
Material properties are not a problem - because they always get reinit'd and resized before they are used.

friedmud commented:
The only issue with (4) is for stateful material properties... it's possible that oversizing the material properties would cause HUGE amounts of memory use.

We really need to get the right number of qps... what about solutions to try to get to the quadrature order earlier?  Like specifying the quadrature order each object needs as a special kind of InputParam... that way we can query all of the object's quadrature order needs _before_ we ever even build them.

rwcarlsen committed:
Modify presized qp data structures to have a large, fixed size.

This is necessary because we want to allow objects to bump the
quadrature order up from their constructors.  But at this stage some
other objects may have already been constructed that pre-size some data
structures based on max qps - these objects will have seen the wrong
(lower) max qps count.  This includes all objects inheriting from
Coupleable or MaterialPropertyInterface.  The default coupled values and
default material property values are just stored on-sight in member-var
data structures.  Rather than working really hard to get a fancy way to
resize these data structures dynamically whenever we bump quadrature
order, we instead just use a large const/fixed qp count for these data
structures as a super duper simple hack.  If needs change in the future
or TheWarehouse gets more universal utilization, we can revisit this
hack/implementation detail.

rwcarlsen commented:
Yaqi seemed to suggest (in an earlier PR/issue discussion about this stuff) that they had use cases where they needed info that wouldn't be available in validParams functions in order to determine the desired quadrature order.  Also I don't think this is a big memory problem.  In general we don't have very large numbers of objects.  And this would add like 1kb to objects that have default coupled var/prop values (not all objects).  Even if you had 10000 of those objects - we are still talking ~10 MB - which is either not a problem because you app doesn't need much memory - or not a problem because you app uses way more memory than that anyway.  Do we know of use cases where people have significantly more than 10,000 objects that use default coupled var values?

rwcarlsen commented:
Also - if it wasn't clear before @friedmud - this doesn't affect the size of material properties - it only affects the size of the "default" material property and "default" coupled value data structures stored on objects themselves - nothing to do with material property storage, etc.  All that stuff remains unchanged using the current/live number of qps for each element.

loganharbour commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -40,7 +40,7 @@ Coupleable::Coupleable(const MooseObject * moose_object, bool nodal, bool is_fv)
                              ? _c_parameters.get<bool>("_neighbor")
                              : false),
-    _coupleable_max_qps(_c_fe_problem.getMaxQps()),
+    _coupleable_max_qps(Moose::constMaxQpsPerElem),
```
I believe this should be
```
std::max(Moose::constMaxQpsPerElem, _c_fe_problem.getMaxQps())
```
so that it'll still work if the default order is higher. Same in MaterialPropertyInterface.

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -4721,6 +4721,10 @@ FEProblemBase::updateMaxQps()
   }
 
   unsigned int max_qpts = getMaxQps();
+  if (max_qpts > Moose::constMaxQpsPerElem)
```
Why not move this into `bumpVolumeQRuleOrder`?

rwcarlsen replied:
Because we don't know what the new max qps is until here.  in bumpVolumeQRuleOrder - we only know what the new order is.

loganharbour commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -40,7 +40,7 @@ Coupleable::Coupleable(const MooseObject * moose_object, bool nodal, bool is_fv)
                              ? _c_parameters.get<bool>("_neighbor")
                              : false),
-    _coupleable_max_qps(_c_fe_problem.getMaxQps()),
+    _coupleable_max_qps(Moose::constMaxQpsPerElem),
```
I believe this should be
```
std::max(Moose::constMaxQpsPerElem, _c_fe_problem.getMaxQps())
```
so that it'll still work if the default order is higher. Same in MaterialPropertyInterface.

rwcarlsen replied:
Yeah - that's a good, simple idea.  But then how do we handle the error message?  We don't want to error in cases where the initial max qps is higher than constMaxQpsPerElem, but we do want to error if it is later bumped to be higher than constMaxQpsPerElem.  I'll come up with something.

loganharbour reviewed (dismissed):
I'm thinking that the const max should _only_ apply if it's bigger than the FE max qps, and we should make it clear in `Moose.h` that this is the case.

Up for discussion, but I would prefer this so that higher global orders will work.

loganharbour reviewed (approved):
👍🏼 comments explaining the unfortunate design choices are clear.

rwcarlsen commented:
@YaqiWang - any last words?

YaqiWang commented:
LGTM. Just make sure Griffin with that commit in my branch indeed works without problems.

rwcarlsen commented:
It worked when I tested it.
